# Capstone Project
### Applied Data Science Capstone by IBM/Coursera

## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Results](#results)
* [Discussion](#discussion)
* [Conclusion](#conclusion)

## Environment

In [1]:
get_ipython().system(u' pip install --upgrade pip')
get_ipython().system(u' pip install beautifulsoup4')
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
!pip install lxml
!pip install html5lib
!pip install requests
!pip install geopy

     |████████████████████████████████| 1.5MB 11.0MB/s eta 0:00:01
  Found existing installation: pip 19.1.1
    Uninstalling pip-19.1.1:
      Successfully uninstalled pip-19.1.1
     |████████████████████████████████| 109 kB 16.7 MB/s eta 0:00:01
     |████████████████████████████████| 5.8 MB 4.6 MB/s eta 0:00:01 0:00:01
     |████████████████████████████████| 104 kB 11.5 MB/s eta 0:00:01


## Introduction: Business Problem <a name="introduction"></a>

In this research, the main purpose will trying to help potential buyers in **Phoenix Metropolitan Area** to looking for suitable area to **living** or **make investments.**

The **Phoenix Metropolitan Area** is a metropolitan area centered on the city of Phoenix, that includes much of the central part of U.S. state of Arizona. Influent by **rapidly rising and unaffordable house price in California**, some people from California considering move to Arizona, and drove the development of the Arizona's economy.

As the reason list in above, we will create a **clustering map** with real estate information in Arizona. According to the venues density, let investors understand better the house pricing in each district.


## Data <a name="data"></a>

Based on the background we discussed above, following are the ways I trying to generate the datas:

* Consider Arizona has large areas are deserts which are uninhabitable or sparsely inhabited. Based on the information from Wikipedia, we only pick the cities that they have over 150,000 inhabitants as following and find the zip codes' list of each cities:

    1. Phoenix
    2. Chandler
    3. Mesa
    4. Scottsdale
    5. Glendale
    6. Gilbert
    7. Tempe
    8. Peoria


* After generate the zip code list from each cities, I got the latitude and longitude data of US zip code from the Opendataoft.com.

* I used Zillow, a real estate info website, to get the median housing price in each zip code.

* From Foursquare API to get the most common venues of given zip of Phoenix Metropolitan Area.

### Web Scrape Zip Data

According the above cities, it needs to find the zip codes' list of each cities. Then, we will try to use a dataframe to record the median house price in each zip code.  

Firstly, we need to the list of zip code from the target cities. We used web scrape tool to scrape the list from the Zip-Code.com.

In [2]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np

# Web scrape the zip data from Zip Code.

cities = ['Phoenix', 'Chandler', 'Mesa', 'Scottsdale', 'Glendale', 'Gilbert', 'Tempe', 'Peoria']
url = 'https://www.zip-codes.com/city/az-{}.asp'.format(cities[0])
url

result = pd.DataFrame()

for i in cities:
    url = 'https://www.zip-codes.com/city/az-{}.asp'.format(i)
    res = requests.get(url)
    soup = BeautifulSoup(res.content,'lxml')
    table = soup.find_all('table', class_="statTable")[0]
    df = pd.read_html(str(table), header = 0)[0]
    df['cities'] = i
#     print(i)
    result = result.append(df)
    
result


ZIP Code      Type    County  Population  \
0   ZIP Code 85001  P.O. Box  Maricopa           0   
1   ZIP Code 85002  P.O. Box  Maricopa           0   
2   ZIP Code 85003  Standard  Maricopa        9369   
3   ZIP Code 85004  Standard  Maricopa        4965   
4   ZIP Code 85005  P.O. Box  Maricopa           0   
5   ZIP Code 85006  Standard  Maricopa       25742   
6   ZIP Code 85007  Standard  Maricopa       14040   
7   ZIP Code 85008  Standard  Maricopa       56145   
8   ZIP Code 85009  Standard  Maricopa       52520   
9   ZIP Code 85010  P.O. Box  Maricopa           0   
10  ZIP Code 85011  P.O. Box  Maricopa           0   
11  ZIP Code 85012  Standard  Maricopa        6390   
12  ZIP Code 85013  Standard  Maricopa       19314   
13  ZIP Code 85014  Standard  Maricopa       24782   
14  ZIP Code 85015  Standard  Maricopa       37644   
15  ZIP Code 85016  Standard  Maricopa       33896   
16  ZIP Code 85017  Standard  Maricopa       38872   
17  ZIP Code 85018  Standard  Maricopa       36065   
18  ZIP Code 85019  Standard  Maricopa       25042   
19  ZIP Code 85020  Standard  Maricopa       32850   
20  ZIP Code 85021  Standard  Maricopa       37743   
21  ZIP Code 85022  Standard  Maricopa       46395   
22  ZIP Code 85023  Standard  Maricopa       31274   
23  ZIP Code 85024  Standard  Maricopa       22898   
24  ZIP Code 85026  Standard  Maricopa           0   
25  ZIP Code 85027  Standard  Maricopa       37140   
26  ZIP Code 85028  Standard  Maricopa       19157   
27  ZIP Code 85029  Standard  Maricopa       43233   
28  ZIP Code 85030  P.O. Box  Maricopa           0   
29  ZIP Code 85031  Standard  Maricopa       30493   
30  ZIP Code 85032  Standard  Maricopa       65485   
31  ZIP Code 85033  Standard  Maricopa       53037   
32  ZIP Code 85034  Standard  Maricopa        5582   
33  ZIP Code 85035  Standard  Maricopa       47486   
34  ZIP Code 85036  P.O. Box  Maricopa           0   
35  ZIP Code 85037  Standard  Maricopa       44771   
36  ZIP Code 85038  P.O. Box  Maricopa           0   
37  ZIP Code 85039  Standard  Maricopa           0   
38  ZIP Code 85040  Standard  Maricopa       29352   
39  ZIP Code 85041  Standard  Maricopa       54947   
40  ZIP Code 85042  Standard  Maricopa       40962   
41  ZIP Code 85043  Standard  Maricopa       30560   
42  ZIP Code 85044  Standard  Maricopa       37993   
43  ZIP Code 85045  Standard  Maricopa        7238   
44  ZIP Code 85046  P.O. Box  Maricopa           0   
45  ZIP Code 85048  Standard  Maricopa       33449   
46  ZIP Code 85050  Standard  Maricopa       25736   
47  ZIP Code 85051  Standard  Maricopa       40375   
48  ZIP Code 85053  Standard  Maricopa       28480   
49  ZIP Code 85054  Standard  Maricopa        5384   
50  ZIP Code 85060  P.O. Box  Maricopa           0   
51  ZIP Code 85061  P.O. Box  Maricopa           0   
52  ZIP Code 85062  P.O. Box  Maricopa           0   
53  ZIP Code 85063  P.O. Box  Maricopa           0   
54  ZIP Code 85064  P.O. Box  Maricopa           0   
55  ZIP Code 85065  Standard  Maricopa           0   
56  ZIP Code 85066  P.O. Box  Maricopa           0   
57  ZIP Code 85067  P.O. Box  Maricopa           0   
58  ZIP Code 85068  P.O. Box  Maricopa           0   
59  ZIP Code 85069  P.O. Box  Maricopa           0   
60  ZIP Code 85070  P.O. Box  Maricopa           0   
61  ZIP Code 85071  P.O. Box  Maricopa           0   
62  ZIP Code 85072  P.O. Box  Maricopa           0   
63  ZIP Code 85073  Standard  Maricopa           0   
64  ZIP Code 85074  P.O. Box  Maricopa           0   
65  ZIP Code 85075  P.O. Box  Maricopa           0   
66  ZIP Code 85076  P.O. Box  Maricopa           0   
67  ZIP Code 85078  P.O. Box  Maricopa           0   
68  ZIP Code 85079  P.O. Box  Maricopa           0   
69  ZIP Code 85080  P.O. Box  Maricopa           0   
70  ZIP Code 85082  P.O. Box  Maricopa           0   
71  ZIP Code 85083  Standard  Maricopa       18104   
72  ZIP Code 85085  Standard  Maricopa       17517   
73  ZIP Code 8

To **avoid the anti-webscraping** from the website, we **save the data to csv** that will help us easy to preprocessing the data.

In [5]:
result.to_csv("PHX_Zip_Code.csv")

In the original data from the web scraper, it has some **unnecessary columns**, like type, county and area codes. It also shows that some zips' population is zero. We will **drop** it in **preprocessing** step.

In [3]:
# Drop the unnessary column
df1 = pd.read_csv("PHX_Zip_Code.csv")
df1 = df1.drop(columns=['Unnamed: 0', 'Type', 'County', 'Area Code(s)'])
df1['ZIP Code'] = df1['ZIP Code'].str.replace('ZIP Code ', '')

# Sort out the zip code which equal to zero
df_sort = df1.loc[df1["Population"] != 0]
df_sort = df_sort.reset_index(drop = True)
df_sort.head()
df_sort[["ZIP Code"]]=df_sort[["ZIP Code"]].astype(int)

df_sort.head()

ZIP Code  Population   cities
0     85003        9369  Phoenix
1     85004        4965  Phoenix
2     85006       25742  Phoenix
3     85007       14040  Phoenix
4     85008       56145  Phoenix

After we got the all zip code of the target cities, we need to find their longitude and latitude to generate 2D map. From **Opendatasoft**, it has a database included all **US zip  code latitude and longitude**. Use the existing list of zip codes, we **merged two table** via pandas.

In [4]:
# Load the database and rename the columns
df_LL = pd.read_csv("PHX_LL.csv")
df_LL = df_LL.drop(columns=['City', 'State', 'Timezone', 'Daylight savings time flag', 'geopoint'])
df_LL = df_LL.rename(columns={'Zip':'ZIP Code'})
df_LL[["ZIP Code"]]=df_LL[["ZIP Code"]].astype(int)

# Merge two dataframe
df3 = pd.merge(df_sort, df_LL, on='ZIP Code', how='left')

#Drop the NAN rows
df3 = df3.dropna()
df3 = df3.reset_index(drop = True)
df3.head()

ZIP Code  Population   cities   Latitude  Longitude
0     85003        9369  Phoenix  33.451143 -112.07838
1     85004        4965  Phoenix  33.451093 -112.07057
2     85006       25742  Phoenix  33.466392 -112.04875
3     85007       14040  Phoenix  33.449995 -112.09024
4     85008       56145  Phoenix  33.466393 -112.00080

Based on the list of the zip codes, we download the **median price** in each zip code from **Zillow**. It's a real estate information website provide house's information of price and history.

In [5]:
df_final = pd.read_csv("PHX_Zip&LL.csv")
df_final = df_final.drop(columns=["Unnamed: 0"])
df_final.head()

ZIP Code  Population   Price   cities   Latitude  Longitude
0     85003        9369  439155  Phoenix  33.451143 -112.07838
1     85004        4965  323521  Phoenix  33.451093 -112.07057
2     85006       25742  287468  Phoenix  33.466392 -112.04875
3     85007       14040  289808  Phoenix  33.449995 -112.09024
4     85008       56145  244292  Phoenix  33.466393 -112.00080

Above are the main data we need to analyze in this case, the **main components** are **zip code, population, price, cities, latitude, longitude**.

### Visualize the Zip Codes in Phoenix Map

In [6]:
# Enable the Enviornment
from geopy.geocoders import Nominatim
import folium
import json
from pandas.io.json import json_normalize
from sklearn.cluster import KMeans

In [7]:
# Get Geographical Coordinate of Phoenix
address = 'Phoenix, AZ'
geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Phoenix are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Phoenix are 33.4484367, -112.0741417.


In [8]:
# Generate the map of Phoenix
map_PHX = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, lng, label in zip(df_final['Latitude'], df_final['Longitude'],df_final['ZIP Code']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_PHX)  

map_PHX

## Methodology <a name="methodology"></a>

After visualize geographic details in maps, I begin to utilized the Foursquare API to explore the neighborhood in each zip code. I set up the limit as 100 venue and the radius 100 meter for each borough from the given zip code longitude and latitude.

### Foursquare API

Now, we have the location info, we will use Foursquare API to acquire the neighborhood which will use for clustering different areas.

In [9]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np

In [10]:
# Access Foursquare API
CLIENT_ID = 'UHDWYG4FTGUOX24P3VTHLNT0ECD3W5KE2YLE03NH1PKJQCNK' # your Foursquare ID
CLIENT_SECRET = 'AQP3UXQ4NEKEPJHDIAKVOVDSJKOJWNHE54RVEIF2FIQYKAEB' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

# Load neighborhood information from each zip
neighborhood_latitude = df_final.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_final.loc[0, 'Longitude'] # neighborhood longitude value

# Set up the limit and radius
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 100 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

# Get neighborhood info in JSON format
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5eb06a241d67cb001bc70533'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Downtown Phoenix',
  'headerFullLocation': 'Downtown Phoenix, Phoenix',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 3,
  'suggestedBounds': {'ne': {'lat': 33.452043000900005,
    'lng': -112.07730333689854},
   'sw': {'lat': 33.4502429991, 'lng': -112.07945666310148}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '5971d45e1f8ed660bbfded85',
       'name': 'The Van Buren',
       'location': {'address': '401 W Van Buren St',
        'lat': 33.451118336974915,
        'lng': -112.07925535943271,
        'labeledLatL

Then, we transfer the format from JSON to dataframe.

In [11]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [12]:
def getNearbyVenues(names, latitudes, longitudes, radius=1500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['ZIP Code', 
                  'Zip Latitude', 
                  'Zip Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [13]:
PHX_venues = getNearbyVenues(names=df_final['ZIP Code'],
                                   latitudes=df_final['Latitude'],
                                   longitudes=df_final['Longitude']
                                  )

85003
85004
85006
85007
85008
85009
85012
85013
85014
85015
85016
85017
85018
85019
85020
85021
85022
85023
85024
85027
85028
85029
85031
85032
85033
85034
85035
85037
85040
85041
85042
85043
85044
85045
85048
85050
85051
85053
85054
85085
85086
85087
85224
85225
85226
85248
85249
85286
85201
85202
85203
85204
85205
85206
85207
85208
85210
85212
85213
85215
85250
85251
85253
85254
85255
85257
85258
85259
85260
85262
85263
85264
85266
85268
85301
85302
85303
85304
85305
85306
85307
85308
85310
85233
85234
85296
85297
85281
85282
85283
85284
85345
85381
85382
85383


In [14]:
# Check how many uniques categories
print(PHX_venues.shape)
PHX_venues.head()
PHX_venues.groupby('ZIP Code').count()
print('There are {} uniques categories.'.format(len(PHX_venues['Venue Category'].unique())))

(3761, 7)
There are 322 uniques categories.


After we got the categories data, we need to **transfer the data to dummy variables**. 

In [15]:
# one hot encoding
PHX_onehot = pd.get_dummies(PHX_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
PHX_onehot['ZIP Code'] = PHX_venues['ZIP Code'] 

# move neighborhood column to the first column
fixed_columns = [PHX_onehot.columns[-1]] + list(PHX_onehot.columns[:-1])
PHX_onehot = PHX_onehot[fixed_columns]

PHX_onehot.head()

ZIP Code  ATM  Accessories Store  Adult Boutique  Advertising Agency  \
0     85003    0                  0               0                   0   
1     85003    0                  0               0                   0   
2     85003    0                  0               0                   0   
3     85003    0                  0               0                   0   
4     85003    0                  0               0                   0   

   African Restaurant  Airport  Airport Service  Airport Terminal  \
0                   0        0                0                 0   
1                   0        0                0                 0   
2                   0        0                0                 0   
3                   0        0                0                 0   
4                   0        0                0                 0   

   Alternative Healer  American Restaurant  Antique Shop  Aquarium  Arcade  \
0                   0                    0             0         0       0   
1                   0                    0             0         0       0   
2                   0                    0             0         0       0   
3                   0                    0             0         0       0   
4                   0                    0             0         0       0   

   Art Gallery  Art Museum  Arts & Crafts Store  Arts & Entertainment  \
0            0           0                    0                     0   
1            0           0                    0                     0   
2            0           0                    0                     0   
3            0           0                    0                     0   
4            0           0                    0                     0   

   Asian Restaurant  Astrologer  Athletics & Sports  Auto Garage  \
0                 0           0                   0            0   
1                 0           0                   0            0   
2                 0           0                   0            0   
3                 0           0                   0            0   
4                 0           0                   0            0   

   Auto Workshop  Automotive Shop  BBQ Joint  Baby Store  Bagel Shop  \
0              0                0          0           0           0   
1              0                0          0           0           0   
2              0                0          0           0           0   
3              0                0          0           0           0   
4              0                0          0           0           0   

   Baggage Locker  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0               0       0     0    0               0                 0   
1               0       0     0    0               0                 0   
2               0       0     0    0               0                 0   
3               0       0     0    0               0                 0   
4               0       0     0    0               0                 0   

   Basketball Court  Basketball Stadium  Bed & Breakfast  Beer Bar  \
0                 0                   0                0         0   
1                 0                   0                0         0   
2                 0                   0                0         0   
3                 0                   0                0         0   
4                 0                   0                0         0   

   Beer Garden  Beer Store  Big Box Store  Bike Shop  Board Shop  \
0            0           0              0          0           0   
1            0           0              0          0           0   
2            0           0              0          0           0   
3            0           0              0          0           0   
4            0           0              0          0           0   

   Boat or Ferry  Bookstore  Boutique  Bowling Alley  Breakfast Spot  Brewery  \
0              0          0    

Next, we need **group rows by neighborhood** and by taking the **mean of the frequency** of occurrence of each category.

In [16]:
PHX_grouped = PHX_onehot.groupby('ZIP Code').mean().reset_index()
PHX_grouped.head()

ZIP Code  ATM  Accessories Store  Adult Boutique  Advertising Agency  \
0     85003  0.0                0.0             0.0                 0.0   
1     85004  0.0                0.0             0.0                 0.0   
2     85006  0.0                0.0             0.0                 0.0   
3     85007  0.0                0.0             0.0                 0.0   
4     85008  0.0                0.0             0.0                 0.0   

   African Restaurant  Airport  Airport Service  Airport Terminal  \
0                 0.0      0.0              0.0               0.0   
1                 0.0      0.0              0.0               0.0   
2                 0.0      0.0              0.0               0.0   
3                 0.0      0.0              0.0               0.0   
4                 0.0      0.0              0.0               0.0   

   Alternative Healer  American Restaurant  Antique Shop  Aquarium  Arcade  \
0            0.000000             0.050000           0.0       0.0    0.01   
1            0.000000             0.050000           0.0       0.0    0.01   
2            0.000000             0.000000           0.0       0.0    0.00   
3            0.000000             0.096774           0.0       0.0    0.00   
4            0.018519             0.000000           0.0       0.0    0.00   

   Art Gallery  Art Museum  Arts & Crafts Store  Arts & Entertainment  \
0     0.060000         0.0             0.000000                   0.0   
1     0.040000         0.0             0.000000                   0.0   
2     0.033898         0.0             0.000000                   0.0   
3     0.032258         0.0             0.032258                   0.0   
4     0.000000         0.0             0.000000                   0.0   

   Asian Restaurant  Astrologer  Athletics & Sports  Auto Garage  \
0               0.0         0.0                 0.0          0.0   
1               0.0         0.0                 0.0          0.0   
2               0.0         0.0                 0.0          0.0   
3               0.0         0.0                 0.0          0.0   
4               0.0         0.0                 0.0          0.0   

   Auto Workshop  Automotive Shop  BBQ Joint  Baby Store  Bagel Shop  \
0            0.0         0.000000        0.0         0.0         0.0   
1            0.0         0.000000        0.0         0.0         0.0   
2            0.0         0.000000        0.0         0.0         0.0   
3            0.0         0.000000        0.0         0.0         0.0   
4            0.0         0.018519        0.0         0.0         0.0   

   Baggage Locker    Bakery      Bank       Bar  Baseball Field  \
0             0.0  0.000000  0.000000  0.030000             0.0   
1             0.0  0.000000  0.000000  0.030000             0.0   
2             0.0  0.016949  0.000000  0.016949             0.0   
3             0.0  0.000000  0.032258  0.000000             0.0   
4             0.0  0.000000  0.000000  0.018519             0.0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Bed & Breakfast  \
0              0.01               0.0                0.01              0.0   
1              0.01               0.0                0.01              0.0   
2              0.00               0.0                0.00              0.0   
3              0.00               0.0                0.00              0.0   
4              0.00               0.0                0.00              0.0   

   Beer Bar  Beer Garden  Beer Store  Big Box Store  Bike Shop  Board Shop  \
0      0.02          0.0         0.0            0.0   0.000000    0.000000   
1      0.02          0.0         0.0            0.0   0.010000    0.000000   
2      0.00          0.0         0.0            0.0   0.016949    0.000000   
3      0.00          0.0         0.0            0.0   0.000000    0.000000   
4      0.00          0.0         0.0            0.0   0.000000    0.018519   

   Boat or Ferry  Bookstore  Boutique  Bowling

In summary of the venues, it has 322 unique categories returned by Foursquare. Now we begin to use the venues to clustering these spaces in Phoenix. We will use **KNN-model** to figure out the **best K**, we should cluster in Phoenix area.

### Find the Best K

We used the KNN model to find the optimum K and will use the K to run K-means algorithm. In training size setting, we set testing size as 20 percent to run unsupervised learning.

In [131]:
import itertools
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.ticker import NullFormatter
import pandas as pd
import numpy as np
import matplotlib.ticker as ticker
from sklearn import preprocessing
from sklearn.neighbors import KNeighborsClassifier
%matplotlib inline

In [154]:
X = PHX_grouped
y = df_final['cities'].values

# Train Test Split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

Train set: (76, 323) (76,)
Test set: (19, 323) (19,)


In [155]:
X = preprocessing.StandardScaler().fit(X).transform(X)

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:1: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  """Entry point for launching an IPython kernel.


In [156]:
# Train Test Split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

Train set: (76, 323) (76,)
Test set: (19, 323) (19,)


In [157]:
# Modeling
from sklearn.neighbors import KNeighborsClassifier
k = 3
#Train Model and Predict
KNN_model = KNeighborsClassifier(n_neighbors=k).fit(X_train, y_train)
KNN_model

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=3, p=2,
           weights='uniform')

In [158]:
import sklearn.metrics as metrics
Ks = 25
mean_acc = np.zeros((Ks-1))
std_acc = np.zeros((Ks-1))
ConfustionMx = [];
for n in range(1,Ks):
    
    #Train Model and Predict  
    neigh = KNeighborsClassifier(n_neighbors = n).fit(X_train,y_train)
    yhat=neigh.predict(X_test)
    mean_acc[n-1] = metrics.accuracy_score(y_test, yhat)

    
    std_acc[n-1]=np.std(yhat==y_test)/np.sqrt(yhat.shape[0])

mean_acc

# Find the biggest accuracy as the best K

print(mean_acc)
print( "The best accuracy was with", mean_acc.max(), "with k=", mean_acc.argmax()+1) 

[0.63157895 0.63157895 0.63157895 0.63157895 0.63157895 0.68421053
 0.68421053 0.68421053 0.63157895 0.63157895 0.63157895 0.63157895
 0.63157895 0.68421053 0.68421053 0.63157895 0.63157895 0.63157895
 0.63157895 0.63157895 0.63157895 0.63157895 0.63157895 0.63157895]
The best accuracy was with 0.6842105263157895 with k= 6


In summary, the result of the KNN model provided the best K for clustering is 8. We will use the optimum K as 6 to run the K-Mean algorithm.

### Table of Top 10 Venues

Then, we want a table of top 10 venues in each zip which we help us easier to show cluster them.

In [159]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [160]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['ZIP Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['ZIP Code'] = PHX_grouped['ZIP Code']

for ind in np.arange(PHX_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(PHX_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

ZIP Code 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0     85003           Coffee Shop           Art Gallery           Pizza Place   
1     85004           Coffee Shop           Pizza Place                 Hotel   
2     85006    Mexican Restaurant            Taco Place        Sandwich Place   
3     85007    Mexican Restaurant     Convenience Store   American Restaurant   
4     85008     Convenience Store  Fast Food Restaurant        Discount Store   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0   American Restaurant           Music Venue                 Hotel   
1   American Restaurant           Art Gallery          Cocktail Bar   
2                  Café     Convenience Store           Art Gallery   
3           Music Venue        History Museum           Art Gallery   
4          Intersection    Mexican Restaurant            Taco Place   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0               Theater                   Bar          Cocktail Bar   
1           Music Venue        Breakfast Spot    Salon / Barbershop   
2           Flower Shop    Seafood Restaurant               Gay Bar   
3            Taco Place      Capitol Building       Thai Restaurant   
4      Sushi Restaurant        Sandwich Place          Burger Joint   

  10th Most Common Venue  
0                    Pub  
1                    Bar  
2           Intersection  
3                Theater  
4          Grocery Store

### Visualization 

In [161]:
# set number of clusters
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

kclusters = 6

PHX_grouped_clustering = PHX_grouped.drop('ZIP Code', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(PHX_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 


array([2, 2, 0, 0, 0, 0, 2, 2, 2, 0], dtype=int32)

In [162]:
# add clustering labels

neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)


PHX_merged = df_final
# PHX_merged = PHX_merged.reset_index(drop=True)

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
PHX_merged = PHX_merged.join(neighborhoods_venues_sorted.set_index('ZIP Code'), on='ZIP Code')

# toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].int()

PHX_merged.head()# check the last columns!

ZIP Code  Population   Price   cities   Latitude  Longitude  \
0     85003        9369  439155  Phoenix  33.451143 -112.07838   
1     85004        4965  323521  Phoenix  33.451093 -112.07057   
2     85006       25742  287468  Phoenix  33.466392 -112.04875   
3     85007       14040  289808  Phoenix  33.449995 -112.09024   
4     85008       56145  244292  Phoenix  33.466393 -112.00080   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               2           Coffee Shop           Art Gallery   
1               2           Coffee Shop           Pizza Place   
2               0    Mexican Restaurant            Taco Place   
3               0    Mexican Restaurant     Convenience Store   
4               0     Convenience Store  Fast Food Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Pizza Place   American Restaurant           Music Venue   
1                 Hotel   American Restaurant           Art Gallery   
2        Sandwich Place                  Café     Convenience Store   
3   American Restaurant           Music Venue        History Museum   
4        Discount Store          Intersection    Mexican Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                 Hotel               Theater                   Bar   
1          Cocktail Bar           Music Venue        Breakfast Spot   
2           Art Gallery           Flower Shop    Seafood Restaurant   
3           Art Gallery            Taco Place      Capitol Building   
4            Taco Place      Sushi Restaurant        Sandwich Place   

  9th Most Common Venue 10th Most Common Venue  
0          Cocktail Bar                    Pub  
1    Salon / Barbershop                    Bar  
2               Gay Bar           Intersection  
3       Thai Restaurant                Theater  
4          Burger Joint          Grocery Store

In [173]:
# create map
import matplotlib.cm as cm
import matplotlib.colors as colors

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(PHX_merged['Latitude'], PHX_merged['Longitude'], PHX_merged['ZIP Code'], PHX_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Results <a name="results"></a>

After the clustering by machine learning with KNN method, you can see Phoenix area has been separate by 6 areas as following,

* Dinning Venues
* Event Venues
* Social Venues
* Construction Venues
* Trail & Park Venues
* Farm Venues

According to different venues, we found that the average price of houses in the central of the Phoenix Metropolitan Area is lower than the peripheral area. Based the result, we would like to classified the house prices as three levels,

* Lower level(Cluster 0): Average Price 286120
* Median Level(Cluster 2): Average Price 392947
* High Level(Cluster 1, 3, 4, 5): Average Price 627426

The result is out of my expectation because the price of central house usually higher than the peripheral area. However, it shows a opposite result in Phoenix. We will do more analysis in discussion section.

### Cluster 0

In [195]:
np.mean(PHX_merged.loc[PHX_merged['Cluster Labels'] == 0, PHX_merged.columns[[2]]])

Price    286120.477273
dtype: float64

In [196]:
PHX_merged.loc[PHX_merged['Cluster Labels'] == 0, PHX_merged.columns[[0] + list(range(2, PHX_merged.shape[1]))]]

ZIP Code   Price      cities   Latitude   Longitude  Cluster Labels  \
2      85006  287468     Phoenix  33.466392 -112.048750               0   
3      85007  289808     Phoenix  33.449995 -112.090240               0   
4      85008  244292     Phoenix  33.466393 -112.000800               0   
5      85009  169682     Phoenix  33.456350 -112.123780               0   
9      85015  227522     Phoenix  33.506890 -112.102620               0   
11     85017  184766     Phoenix  33.514092 -112.122850               0   
13     85019  193564     Phoenix  33.511992 -112.141700               0   
14     85020  276980     Phoenix  33.563663 -112.055190               0   
15     85021  298364     Phoenix  33.559437 -112.092810               0   
16     85022  303365     Phoenix  33.630497 -112.052100               0   
17     85023  275752     Phoenix  33.638271 -112.093410               0   
18     85024  316152     Phoenix  33.675564 -112.047170               0   
20     85028  423219     Phoenix  33.585037 -112.008150               0   
21     85029  222900     Phoenix  33.598841 -112.120230               0   
22     85031  186237     Phoenix  33.493496 -112.171080               0   
23     85032  292044     Phoenix  33.624140 -112.004160               0   
24     85033  194637     Phoenix  33.496091 -112.214050               0   
26     85035  194820     Phoenix  33.472492 -112.187500               0   
27     85037  215429     Phoenix  33.493006 -112.255390               0   
28     85040  190646     Phoenix  33.393323 -112.028570               0   
29     85041  229684     Phoenix  33.386995 -112.100760               0   
31     85043  224360     Phoenix  33.435405 -112.200090               0   
34     85048  384868     Phoenix  33.302098 -112.024770               0   
35     85050  380496     Phoenix  33.683603 -111.992710               0   
36     85051  210959     Phoenix  33.559783 -112.133610               0   
40     85086  388039     Phoenix  33.831866 -112.063390               0   
41     85087  401421     Phoenix  33.921493 -112.099920               0   
43     85225  290770    Chandler  33.316349 -111.830640               0   
48     85201  232568        Mesa  33.432177 -111.847010               0   
51     85204  248926        Mesa  33.400127 -111.785940               0   
52     85205  265751        Mesa  33.426932 -111.716860               0   
55     85208  247682        Mesa  33.393967 -111.643020               0   
58     85213  359578        Mesa  33.434192 -111.768920               0   
59     85215  326183        Mesa  33.479868 -111.693500               0   
70     85263  557854  Scottsdale  33.722594 -111.679150               0   
74     85301  181134    Glendale  33.534128 -112.176700               0   
76     85303  230699    Glendale  33.528166 -112.221890               0   
77     85304  254918    Glendale  33.594069 -112.175310               0   
83     85233  339136     Gilbert  33.348593 -111.808330               0   
85     85296  346939     Gilbert  33.319722 -111.760780               0   
86     85297  376191     Gilbert  33.272322 -111.714171               0   
90     85284  473621       Tempe  33.338333 -111.929620               0   
91     85345  234689      Peoria  33.573602 -112.239840               0   
94     85383  415188      Peoria  33.700837 -112.244827               0   

   1st Most Common Venue   2nd Most Common Venue   3rd Most Common Venue  \
2     Mexican Restaurant              Taco Place          Sandwich Place   
3     Mexican Restaurant       Convenience Store     American Restaurant   
4      Convenience Store    Fast Food Restaurant          Discount Store   
5     Mexican Restaurant       Convenience Store    Fast Food Restaurant   
9      Convenience Store             Pizza Place      Mexican Restaurant   
11           Pizza Place    Fast Food Restaurant              Taco Place   
13    Mexican Restaurant       Convenience Store             Pizza Place   
14    Mexican Restau

### Cluster 1

In [197]:
np.mean(PHX_merged.loc[PHX_merged['Cluster Labels'] == 1, PHX_merged.columns[[2]]])

Price    784550.0
dtype: float64

In [181]:
PHX_merged.loc[PHX_merged['Cluster Labels'] == 1, PHX_merged.columns[[0] + list(range(2, PHX_merged.shape[1]))]]

ZIP Code   Price      cities   Latitude  Longitude  Cluster Labels  \
72     85266  784550  Scottsdale  33.276539 -112.18717               1   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
72            Food Truck            Comic Shop         Event Service   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
72           Event Space            Eye Doctor           Fabric Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
72    Falafel Restaurant                  Farm        Farmers Market   

   10th Most Common Venue  
72   Fast Food Restaurant

### Cluster 2

In [198]:
np.mean(PHX_merged.loc[PHX_merged['Cluster Labels'] == 2, PHX_merged.columns[[2]]])

Price    392946.976744
dtype: float64

In [182]:
PHX_merged.loc[PHX_merged['Cluster Labels'] == 2, PHX_merged.columns[[0] + list(range(2, PHX_merged.shape[1]))]]

ZIP Code    Price      cities   Latitude   Longitude  Cluster Labels  \
0      85003   439155     Phoenix  33.451143 -112.078380               2   
1      85004   323521     Phoenix  33.451093 -112.070570               2   
6      85012   507832     Phoenix  33.505251 -112.070300               2   
7      85013   344588     Phoenix  33.507110 -112.084830               2   
8      85014   311910     Phoenix  33.511325 -112.057110               2   
10     85016   358155     Phoenix  33.503742 -112.032350               2   
12     85018   577671     Phoenix  33.498076 -111.986030               2   
19     85027   250155     Phoenix  33.698731 -112.114290               2   
25     85034   182158     Phoenix  33.437772 -112.028100               2   
30     85042   243565     Phoenix  33.381234 -112.027663               2   
32     85044   322829     Phoenix  33.333591 -111.992460               2   
37     85053   244956     Phoenix  33.629911 -112.131220               2   
38     85054   436419     Phoenix  33.689558 -111.964630               2   
42     85224   298355    Chandler  33.328951 -111.873500               2   
44     85226   337926    Chandler  33.312208 -111.931770               2   
45     85248   387457    Chandler  33.239097 -111.863550               2   
46     85249   436176    Chandler  33.237229 -111.800400               2   
49     85202   260568        Mesa  33.381097 -111.875600               2   
50     85203   285664        Mesa  33.437493 -111.806000               2   
53     85206   258089        Mesa  33.400306 -111.724520               2   
54     85207   363456        Mesa  33.436767 -111.643380               2   
56     85210   242381        Mesa  33.391233 -111.841290               2   
60     85250   400640  Scottsdale  33.521433 -111.909440               2   
61     85251   375171  Scottsdale  33.494152 -111.920790               2   
62     85253  1650254  Scottsdale  33.544596 -111.956450               2   
63     85254   541963  Scottsdale  33.617504 -111.954160               2   
64     85255   765894  Scottsdale  33.697268 -111.883210               2   
65     85257   352380  Scottsdale  33.467347 -111.917410               2   
66     85258   545491  Scottsdale  33.566635 -111.893370               2   
67     85259   716579  Scottsdale  33.596995 -111.826280               2   
68     85260   477677  Scottsdale  33.606004 -111.886990               2   
73     85268   443222  Scottsdale  33.604811 -111.728520               2   
75     85302   245639    Glendale  33.568143 -112.176990               2   
78     85305   292717    Glendale  33.529373 -112.252250               2   
79     85306   259164    Glendale  33.624546 -112.176700               2   
80     85307   239201    Glendale  33.529314 -112.306070               2   
81     85308   314033    Glendale  33.657722 -112.178250               2   
84     85234   356107     Gilbert  33.363440 -111.750930               2   
87     85281   286432       Tempe  33.426885 -111.927330               2   
88     85282   313303       Tempe  33.394067 -111.928730               2   
89     85283   307078       Tempe  33.365951 -111.931550               2   
92     85381   298914      Peoria  33.608640 -112.227910               2   
93     85382   301875      Peoria  33.689869 -112.248380               2   

   1st Most Common Venue       2nd Most Common Venue   3rd Most Common Venue  \
0            Coffee Shop                 Art Gallery             Pizza Place   
1            Coffee Shop                 Pizza Place                   Hotel   
6            Pizza Place              Ice Cream Shop     American Restaurant   
7            Coffee Shop          Mexican Restaurant             Pizza Place   
8     Mexican Restaurant                 Pizza Place          Sandwich Place   
10   American Restaurant               Grocery Store      Mexican Restaurant   
12           Coffee Shop                 Pizza Place     American Restaurant   
19                 

### Cluster 3

In [199]:
np.mean(PHX_merged.loc[PHX_merged['Cluster Labels'] == 3, PHX_merged.columns[[2]]])

Price    859155.0
dtype: float64

In [200]:
PHX_merged.loc[PHX_merged['Cluster Labels'] == 3, PHX_merged.columns[[0] + list(range(2, PHX_merged.shape[1]))]]

ZIP Code   Price      cities   Latitude  Longitude  Cluster Labels  \
71     85264  859155  Scottsdale  33.617643 -111.67554               3   

         1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
71  Construction & Landscaping    Athletics & Sports     Convenience Store   

         4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
71  Financial or Legal Service           Event Space            Eye Doctor   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
71           Fabric Shop    Falafel Restaurant                  Farm   

   10th Most Common Venue  
71         Farmers Market

### Cluster 4

In [201]:
np.mean(PHX_merged.loc[PHX_merged['Cluster Labels'] == 4, PHX_merged.columns[[2]]])

Price    505586.2
dtype: float64

In [202]:
PHX_merged.loc[PHX_merged['Cluster Labels'] == 4, PHX_merged.columns[[0] + list(range(2, PHX_merged.shape[1]))]]

ZIP Code   Price      cities   Latitude  Longitude  Cluster Labels  \
33     85045  422175     Phoenix  33.299740 -112.09812               4   
39     85085  410482     Phoenix  33.777760 -112.07945               4   
47     85286  407031    Chandler  33.850506 -112.05927               4   
69     85262  940022  Scottsdale  33.798581 -111.87080               4   
82     85310  348221    Glendale  33.701173 -112.17614               4   

         1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
33                       Trail           Golf Course           Coffee Shop   
39                       Trail           Yoga Studio        Farmers Market   
47  Construction & Landscaping                  Café                 Trail   
69                 Golf Course                  Food                   Gym   
82                       Trail          Intersection                  Park   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
33          Intersection        Farmers Market    English Restaurant   
39    English Restaurant         Event Service           Event Space   
47        Farmers Market         Event Service           Event Space   
69                 Trail    Falafel Restaurant     Electronics Store   
82           Golf Course        Sandwich Place          Home Service   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
33         Event Service           Event Space            Eye Doctor   
39            Eye Doctor           Fabric Shop    Falafel Restaurant   
47            Eye Doctor           Fabric Shop    Falafel Restaurant   
69    English Restaurant         Event Service           Event Space   
82                Lawyer              Pharmacy           Event Space   

   10th Most Common Venue  
33            Fabric Shop  
39                   Farm  
47                   Farm  
69             Eye Doctor  
82            Flower Shop

### Cluster 5

In [203]:
np.mean(PHX_merged.loc[PHX_merged['Cluster Labels'] == 5, PHX_merged.columns[[2]]])

Price    360416.0
dtype: float64

In [185]:
PHX_merged.loc[PHX_merged['Cluster Labels'] == 5, PHX_merged.columns[[0] + list(range(2, PHX_merged.shape[1]))]]

ZIP Code   Price cities  Latitude  Longitude  Cluster Labels  \
57     85212  360416   Mesa  33.35035 -111.64535               5   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
57                  Farm         Garden Center           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
57  Fast Food Restaurant         Event Service           Event Space   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
57            Eye Doctor           Fabric Shop    Falafel Restaurant   

   10th Most Common Venue  
57         Farmers Market

Average price of cluster 1, 3, 4, 5

In [205]:
(np.mean(PHX_merged.loc[PHX_merged['Cluster Labels'] == 1, PHX_merged.columns[[2]]])+np.mean(PHX_merged.loc[PHX_merged['Cluster Labels'] == 3, PHX_merged.columns[[2]]])+\
np.mean(PHX_merged.loc[PHX_merged['Cluster Labels'] == 4, PHX_merged.columns[[2]]])+np.mean(PHX_merged.loc[PHX_merged['Cluster Labels'] == 5, PHX_merged.columns[[2]]]))/4

Price    627426.8
dtype: float64

## Discussion <a name="discussion"></a>

As the result showing, Phoenix area has three levels of housing price, and the house's allocations are different than our general thinking. The central average price of the house are lower than peripheral area. We think the reason cause the result because we compared with total sale price rather than unit sale price. In peripheral area, the houses usually have a larger area than the houses in central area. Although the peripheral house have higher sale price, their unit sale price are lower than the central houses. 

To improve the machine learning model, I will add the unit sale price and monthly rent rate as the addition variables. It will help investors has a more clear idea to make the investment decision.

## Conclusion <a name="conclusion"></a>

Based on the analysis the price of Phoenix area, we recommend potential investor focus on the houses in cluster 0 and cluster 2. Their sale price and more attractable than peripheral houses, not only on sale prices, but also their population are higher than peripheral. It would help investors to reduce the risk of the investment, because higher population will effectively reduce the vacant rate of the renting houses. 